jscomp/ml/ast_untagged_variants.ml

```diff
@@ -117,7 +117,7 @@ let type_is_builtin_object (t : Types.type_expr) =
   match t.desc with
   | Tconstr (path, _, _) ->
     let name = Path.name path in
-    name = "Js.Dict.t" || name = "Js_dict.t"
+    name = "Js.Dict.t" || name = "Js_dict.t" || name = "Js.Re.t" || name = "RescriptCore.Re.t"
```

Comment:
For the cases when `Js`/`Belt`/`RescriptCore` opened globally:
```suggestion
    name = "Js.Dict.t" || name = "Js_dict.t" || name = "Dict.t" || "RescriptCore.Dict.t" || name = "Js.Re.t" || name = "Re.t" || name = "RescriptCore.Re.t"
```

Reply:
That's not how it works. Internally, you still get the full path.

Reply:
<img width="389" alt="Screenshot 2023-06-08 at 20 36 03" src="https://github.com/rescript-lang/rescript-compiler/assets/7965335/4104857f-6602-4aa4-a956-781b420db90f">

Reply:
I'm talking about the case when there's a custom implementation of a module from stdlib.
<img width="1123" alt="image" src="https://github.com/rescript-lang/rescript-compiler/assets/49292466/723ff3b5-3df3-48a8-9726-854097c1a9cb">

Reply:
Got it. That is another topic, which is pretty interesting, and worth discussing in a separate issue.

Quick thoughts: assuming it's what it is just based on name sounds risky, and perhaps not even necessary.

Topic: how to represent that a type is an object type.
1. For untagged variants, it's useful to express more patterns
2. For options, it helps as they can be represented unboxed
3. For the language: the current mechanism `{..}` is both obscure and not very useful/flexible.

For points 1 and 2, one can already recognise when a type used is defined as an object type (unless I'm misremembering -- check).
For abstract types, one could e.g. use an annotation `@object`.

For 3, perhaps one could use a new built-in type `object<_>` with some operations on it. E.g. creation will return `object<'a>` for an arbitrary `'a`.
There has to be a mechanism to cast actual objects into the type `object<...>`.
The `Dict` module might look very different if this is used.
So having an `object<...>` type is only one possibility. But if one embraces this possibility then the case of the abstract types would work like this:
```
type someSecretType
type someAbstractObject = object<someSecretType>
```
without the need for any `@object` annotation.